src/hotspot/share/opto/vectorization.hpp

```diff
@@ -1323,6 +1323,350 @@ class AlignmentSolver {
+    assert(vector_length > 0 && vector_width > 0, "must have nonzero size");
+    return VTransformApplyStatus(n, vector_length, vector_width);
+  }
+
+  static VTransformApplyStatus make_empty() {
+    return VTransformApplyStatus(nullptr, 0, 0);
+  }
+
+  Node* node() const { return _node; }
+  uint vector_length() const { return _vector_length; }
+  uint vector_width() const { return _vector_width; }
+  NOT_PRODUCT( void trace(VTransformNode* vtn) const; )
+};
+
+class VTransformGraph : public StackObj {
+private:
+  const VLoopAnalyzer& _vloop_analyzer;
+  const VLoop& _vloop;
+
+  // Everything in the graph is allocated from this arena, including all vtnodes.
+  Arena _arena;
+
+  VTransformNodeIDX _next_idx;
+  GrowableArray<VTransformNode*> _vtnodes;
+
+  // Schedule (linearization) of the graph. We use this to reorder the memory graph
+  // before inserting vector operations.
+  GrowableArray<VTransformNode*> _schedule;
+
+  // Memory reference, and the alignment width (aw) for which we align the main-loop,
+  // by adjusting the pre-loop limit.
+  MemNode const* _mem_ref_for_main_loop_alignment;
+  int _aw_for_main_loop_alignment;
+
+#ifndef PRODUCT
+  bool _is_trace_rejections;
+  bool _is_trace_align_vector;
+  bool _is_trace_info;
+  bool _is_trace_verbose;
+#endif
+
+public:
+  VTransformGraph(const VLoopAnalyzer& vloop_analyzer,
+                  MemNode const* mem_ref_for_main_loop_alignment,
+                  int aw_for_main_loop_alignment
+                  NOT_PRODUCT( COMMA const bool is_trace_rejections)
+                  NOT_PRODUCT( COMMA const bool is_trace_align_vector)
+                  NOT_PRODUCT( COMMA const bool is_trace_info)
+                  ) :
+    _vloop_analyzer(vloop_analyzer),
+    _vloop(vloop_analyzer.vloop()),
+    _arena(mtCompiler),
+    _next_idx(0),
+    _vtnodes(&_arena, _vloop.estimated_body_length(), 0, nullptr),
+    _schedule(&_arena, _vloop.estimated_body_length(), 0, nullptr),
+    _mem_ref_for_main_loop_alignment(mem_ref_for_main_loop_alignment),
+    _aw_for_main_loop_alignment(aw_for_main_loop_alignment)
+    NOT_PRODUCT( COMMA _is_trace_rejections(is_trace_rejections) )
+    NOT_PRODUCT( COMMA _is_trace_align_vector(is_trace_align_vector) )
+    NOT_PRODUCT( COMMA _is_trace_info(is_trace_info) )
+  {
+#ifndef PRODUCT
+    bool is_trace     = _vloop.vtrace().is_trace(TraceAutoVectorizationTag::VTRANSFORM);
+    _is_trace_verbose = _vloop.vtrace().is_trace(TraceAutoVectorizationTag::ALL);
+    _is_trace_rejections   |= is_trace || _is_trace_verbose;
+    _is_trace_align_vector |= is_trace || _is_trace_verbose;
+    _is_trace_info         |= is_trace || _is_trace_verbose;
+#endif
+  }
+
+  const VLoopAnalyzer& vloop_analyzer() const { return _vloop_analyzer; }
+  Arena* arena() { return &_arena; }
+  VTransformNodeIDX new_idx() { return _next_idx++; }
+  void add_vtnode(VTransformNode* vtnode);
+
+  bool schedule();
+  void apply();
```

Comment:
Note: these are the public methods. In the future, we can add more, like `cost()` from the cost-model, or `optimize()` if we have optimizations, or maybe `if_convert()`.

Reply:
I think `VTransformGraph::apply` is much better than `VTransformGraph::transform_vtnodes`. Because I want to say that we are "applying" or "executing" the transform, which operates on the C2 graph. We are thus transforming some C2 nodes into some other C2 nodes. `transform_vtnodes` suggests that we are transforming the vtnodes, but "from what into what?" would be my intuitive question if I read this.

Would `apply_to_C2_nodes` or `apply_to_C2_graph` be better for you? Or do you have any other ideas?